File: backend/src/main/java/com/axora/backend/service/UserService.java
package com.axora.backend.service;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.axora.backend.dto.user.ChangePasswordRequest;
import com.axora.backend.dto.user.UserRequest;
import com.axora.backend.dto.user.UserResponse;
import com.axora.backend.entity.User;
import com.axora.backend.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class UserService implements UserDetailsService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return userRepository.findByEmail(username)
                .orElseThrow(() -> new UsernameNotFoundException("Kullanıcı bulunamadı: " + username));
    }

    public List<UserResponse> getAllUsers() {
        return userRepository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public UserResponse getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı"));
        return mapToResponse(user);
    }

    public UserResponse createUser(UserRequest request) {
        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(request.getRole())
                .active(true)
                .build();

        return mapToResponse(userRepository.save(user));
    }

    public UserResponse updateUser(Long id, UserRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı"));

        user.setName(request.getName());
        user.setEmail(request.getEmail());
        if (request.getPassword() != null && !request.getPassword().isEmpty()) {
            user.setPassword(passwordEncoder.encode(request.getPassword()));
        }
        user.setRole(request.getRole());
        user.setActive(request.isActive());

        return mapToResponse(userRepository.save(user));
    }

    public void deleteUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı"));
        user.setActive(false);
        userRepository.save(user);
    }

    public User getCurrentUser() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı"));
    }

    // Profil yönetimi metodları
    public UserResponse getCurrentUserProfile() {
        return mapToResponse(getCurrentUser());
    }

    public UserResponse updateCurrentUserProfile(UserRequest request) {
        User currentUser = getCurrentUser();
        
        // Email değişikliği için kontrol
        if (!currentUser.getEmail().equals(request.getEmail()) && 
            userRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new RuntimeException("Bu e-posta adresi zaten kullanılıyor");
        }

        currentUser.setName(request.getName());
        currentUser.setEmail(request.getEmail());

        return mapToResponse(userRepository.save(currentUser));
    }

    public void changePassword(ChangePasswordRequest request) {
        User currentUser = getCurrentUser();

        // Mevcut şifre kontrolü
        if (!passwordEncoder.matches(request.getCurrentPassword(), currentUser.getPassword())) {
            throw new RuntimeException("Mevcut şifre yanlış");
        }

        currentUser.setPassword(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(currentUser);
    }

    private UserResponse mapToResponse(User user) {
        return UserResponse.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole())
                .active(user.isActive())
                .createdAt(user.getCreatedAt())
                .updatedAt(user.getUpdatedAt())
                .build();
    }
} 

File: frontend/src/services/userService.ts
import axios from "@/lib/axios"
import { User } from "@/types/user"

export const userService = {
  getUsers: async () => {
    const response = await axios.get<User[]>("/api/users")
    return response.data
  },

  getUserById: async (id: number) => {
    const response = await axios.get<User>(`/api/users/${id}`)
    return response.data
  },

  createUser: async (data: Omit<User, "id" | "createdAt" | "updatedAt">) => {
    const response = await axios.post<User>("/api/users", data)
    return response.data
  },

  updateUser: async (id: number, data: Partial<Omit<User, "id" | "createdAt" | "updatedAt">>) => {
    const response = await axios.put<User>(`/api/users/${id}`, data)
    return response.data
  },

  deleteUser: async (id: number) => {
    const response = await axios.delete<void>(`/api/users/${id}`)
    return response.data
  },

  // Profil yönetimi için yeni metodlar
  getProfile: async () => {
    const response = await axios.get<User>("/api/users/profile")
    return response.data
  },

  updateProfile: async (data: { name: string; email: string }) => {
    const response = await axios.put<User>("/api/users/profile", data)
    return response.data
  },

  changePassword: async (data: { currentPassword: string; newPassword: string }) => {
    const response = await axios.put<void>("/api/users/profile/password", data)
    return response.data
  },
} 